src/dbnode/encoding/istream.go

```diff
@@ -21,169 +21,119 @@
-	if is.err != nil {
-		return 0, is.err
-	}
-	if is.remaining == 0 {
-		if err := is.readByteFromStream(); err != nil {
-			return 0, err
-		}
-	}
-	return Bit(is.consumeBuffer(1)), nil
+// NewIStream creates a new IStream
+func NewIStream(reader64 xio.Reader64) *IStream {
+	return &IStream{r: reader64}
 }
 
-func (is *istream) Read(b []byte) (int, error) {
-	if is.remaining == 0 {
-		// Optimized path for when the iterator is already aligned on a byte boundary. Avoids
-		// all the bit manipulation and ReadByte() function calls.
-		// Use ReadFull because the bufferedReader may not return the requested number of bytes.
-		return io.ReadFull(is.r, b)
-	}
-
-	var (
-		i   int
-		err error
-	)
-
+// Read reads len(b) bytes.
+func (is *IStream) Read(b []byte) (int, error) {
+	var i int
 	for ; i < len(b); i++ {
-		b[i], err = is.ReadByte()
+		res, err := is.ReadBits(8)
 		if err != nil {
 			return i, err
 		}
+		b[i] = byte(res)
 	}
 	return i, nil
 }
 
-func (is *istream) ReadByte() (byte, error) {
-	if is.err != nil {
-		return 0, is.err
-	}
-	remaining := is.remaining
-	res := is.consumeBuffer(remaining)
-	if remaining == 8 {
-		return res, nil
-	}
-	if err := is.readByteFromStream(); err != nil {
-		return 0, err
-	}
-	res = (res << uint(8-remaining)) | is.consumeBuffer(8-remaining)
-	return res, nil
+// ReadByte reads the next Byte.
+func (is *IStream) ReadByte() (byte, error) {
+	res, err := is.ReadBits(8)
+	return byte(res), err
+}
+
+// ReadBit reads the next Bit.
+func (is *IStream) ReadBit() (Bit, error) {
+	res, err := is.ReadBits(1)
+	return Bit(res), err
 }
 
-func (is *istream) ReadBits(numBits uint) (uint64, error) {
+// ReadBits reads the next Bits.
+func (is *IStream) ReadBits(numBits uint8) (uint64, error) {
 	if is.err != nil {
 		return 0, is.err
 	}
-	var res uint64
-	numBytes := numBits / 8
-	if numBytes > 0 {
-		// Use Read call rather than individual ReadByte calls since it has
-		// optimized path for when the iterator is aligned on a byte boundary.
-		bytes := is.buffer[0:numBytes]
-		_, err := is.Read(bytes)
-		if err != nil {
-			return 0, err
-		}
-		for _, b := range bytes {
-			res = (res << 8) | uint64(b)
-		}
+	if numBits <= is.remaining {
+		return is.consumeBuffer(numBits), nil
 	}
-
-	numBits = numBits % 8
-	for numBits > 0 {
-		// This is equivalent to calling is.ReadBit() in a loop but some manual inlining
-		// has been performed to optimize this loop as its heavily used in the hot path.
-		if is.remaining == 0 {
-			if err := is.readByteFromStream(); err != nil {
-				return 0, err
-			}
-		}
-
-		numToRead := numBits
-		if is.remaining < numToRead {
-			numToRead = is.remaining
-		}
-		bits := is.current >> (8 - numToRead)
-		is.current <<= numToRead
-		is.remaining -= numToRead
-		res = (res << uint64(numToRead)) | uint64(bits)
-		numBits -= numToRead
+	res := readBitsInWord(is.current, numBits)
+	bitsNeeded := numBits - is.remaining
+	if err := is.readWordFromStream(); err != nil {
```

Comment:
Not following this logic here... shouldn't we check if there are remaining bits to be reading before reading another word from the stream? 

Might be worth adding some comments here around the logic.

Reply:
There is such check with an early return few lines above:
```go
	if numBits <= is.remaining {
		// Have enough bits buffered.
		return is.consumeBuffer(numBits), nil
	}
```
I'll add some comments for clarity.